Engine_Revit_UI/Query/CeilingPattern.cs

```diff
@@ -0,0 +1,203 @@
+        {
+            CeilingType ceilingType = ceiling.Document.GetElement(ceiling.GetTypeId()) as CeilingType;
+            CompoundStructure comStruct = ceilingType.GetCompoundStructure();
+
+            Material material = null;
+            if (comStruct != null && comStruct.GetLayers().Count > 0)
+                material = ceiling.Document.GetElement(comStruct.GetLayers().Last().MaterialId) as Material;
+            else
+            {
+                List<ElementId> materialIds = ceiling.GetMaterialIds(false).ToList();
+                if (materialIds.Count == 0)
+                {
+                    BH.Engine.Reflection.Compute.RecordError("Ceiling patterns could not be pulled for ceiling element with the ID " + ceiling.Id);
+                    return new List<oM.Geometry.Line>();
+                }
+                material = ceiling.Document.GetElement(materialIds[0]) as Material;
+            }
+
+            List<oM.Geometry.Line> result = new List<oM.Geometry.Line>();
+            if (surface == null)
+            {
+                //This would need to be extended to take openings from Values into account
+                foreach (PlanarSurface srf in ceiling.PlanarSurfaceDictionary(false, settings).Keys)
+                {
+                    result.AddRange(material.CeilingPattern(srf, settings));
+                }
+            }
+            else
+                result.AddRange(material.CeilingPattern(surface, settings));
+
+            return result;
+        }
+
+        /***************************************************/
+
+        public static List<BH.oM.Geometry.Line> CeilingPattern(this Material revitMaterial, PlanarSurface surface, RevitSettings settings)
+        {
+            if (surface.InternalBoundaries.Count > 0)
+                BH.Engine.Reflection.Compute.RecordWarning("Currently ceiling openings are not taken into account when generating ceiling patterns.");
+
+            BoundingBox box = surface.IBounds();
+            double z = surface.ExternalBoundary.IControlPoints().Max(x => x.Z);
+            double yLength = box.Max.Y - box.Min.Y;
+            double xLength = box.Max.X - box.Min.X;
+            double maxBoxLength = box.Min.Distance(box.Max) * 2;
+
+            BH.oM.Geometry.Line leftLine = new oM.Geometry.Line
+            {
+                Start = new oM.Geometry.Point { X = box.Min.X, Y = box.Min.Y, Z = z },
+                End = new oM.Geometry.Point { X = box.Min.X, Y = box.Max.Y, Z = z },
+            };
+
+            BH.oM.Geometry.Line rightLine = new oM.Geometry.Line
+            {
+                Start = new oM.Geometry.Point { X = box.Max.X, Y = box.Min.Y, Z = z },
+                End = new oM.Geometry.Point { X = box.Max.X, Y = box.Max.Y, Z = z },
+            };
+
+            List<BH.oM.Geometry.Line> boundarySegments = surface.ExternalBoundary.ICollapseToPolyline(BH.oM.Geometry.Tolerance.Angle).ISubParts().Select(x => x as BH.oM.Geometry.Line).ToList();
```

Comment:
This can be chopped down to `List<BH.oM.Geometry.Line> boundarySegments = surface.ExternalBoundary.ICollapseToPolyline(BH.oM.Geometry.Tolerance.Angle).SubParts().ToList();`